class java.lang.ref.FinalizerReference$Sentinel extends java.lang.Object
{
    boolean finalized;

    private void <init>()
    {
        java.lang.ref.FinalizerReference$Sentinel r0;

        r0 := @this: java.lang.ref.FinalizerReference$Sentinel;

        specialinvoke r0.<java.lang.Object: void <init>()>();

        r0.<java.lang.ref.FinalizerReference$Sentinel: boolean finalized> = 0;

        return;
    }

    void <init>(java.lang.ref.FinalizerReference$1)
    {
        java.lang.ref.FinalizerReference$Sentinel r0;
        java.lang.ref.FinalizerReference$1 r1;

        r0 := @this: java.lang.ref.FinalizerReference$Sentinel;

        r1 := @parameter0: java.lang.ref.FinalizerReference$1;

        specialinvoke r0.<java.lang.ref.FinalizerReference$Sentinel: void <init>()>();

        return;
    }

    void awaitFinalization() throws java.lang.InterruptedException
    {
        java.lang.ref.FinalizerReference$Sentinel r0;
        java.lang.Throwable r1, $r2;
        boolean $z0;

        r0 := @this: java.lang.ref.FinalizerReference$Sentinel;

        entermonitor r0;

     label1:
        $z0 = r0.<java.lang.ref.FinalizerReference$Sentinel: boolean finalized>;

        if $z0 != 0 goto label4;

        virtualinvoke r0.<java.lang.Object: void wait()>();

     label2:
        goto label1;

     label3:
        $r2 := @caughtexception;

        r1 = $r2;

        exitmonitor r0;

        throw r1;

     label4:
        exitmonitor r0;

        return;

        catch java.lang.Throwable from label1 to label2 with label3;
    }

    protected void finalize() throws java.lang.Throwable
    {
        java.lang.ref.FinalizerReference$Sentinel r0;
        java.lang.Throwable r1, $r3;
        boolean $z0;
        java.lang.AssertionError $r2;

        r0 := @this: java.lang.ref.FinalizerReference$Sentinel;

        entermonitor r0;

     label1:
        $z0 = r0.<java.lang.ref.FinalizerReference$Sentinel: boolean finalized>;

        if $z0 == 0 goto label3;

        $r2 = new java.lang.AssertionError;

        specialinvoke $r2.<java.lang.AssertionError: void <init>()>();

        throw $r2;

     label2:
        $r3 := @caughtexception;

        r1 = $r3;

        exitmonitor r0;

        throw r1;

     label3:
        r0.<java.lang.ref.FinalizerReference$Sentinel: boolean finalized> = 1;

        virtualinvoke r0.<java.lang.Object: void notifyAll()>();

     label4:
        exitmonitor r0;

        return;

        catch java.lang.Throwable from label1 to label2 with label2;
        catch java.lang.Throwable from label3 to label4 with label2;
    }
}
